### PR TITLE
Migrate the cmr-client package to TypeScript

### PR DESCRIPTION
**Pull request by yjpa7145**
**Summary:** Migrate the cmr-client package to TypeScript
Addresses [CUMULUS-1927: Migrate the @cumulus/cmr-client package to Typescript](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1927)

## Changes

* Migrate the `@cumulus/cmr-client` package to TypeScript

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: packages/cmr-client/package.json [hunk 1]
```diff
@@ -35,9 +35,12 @@
   "dependencies": {
     "@cumulus/aws-client": "2.0.1",
     "@cumulus/logger": "2.0.1",
-    "got": "^9.6.0",
+    "got": "^11.7.0",
     "lodash": "^4.17.15",
     "public-ip": "^3.0.0",
     "xml2js": "^0.4.19"
+  },
+  "devDependencies": {
```
Comment (markdboyd) [inline packages/cmr-client/package.json:43]: Tangential question: why are these dev dependencies not moved to the top level?

File: packages/cmr-client/src/CMR.ts [hunk 2]
```diff
@@ -36,20 +34,26 @@ const userIpAddress = () =>
  *
  * @private
  */
-async function updateToken(cmrProvider, clientId, username, password) {
-  // if (!cmrProvider) throw new Error('cmrProvider is required.');
-  // if (!clientId) throw new Error('clientId is required.');
-  // if (!username) throw new Error('username is required.');
-  // if (!password) throw new Error('password is required.');
+async function updateToken(
+  cmrProvider: string,
```
Comment (markdboyd) [inline packages/cmr-client/src/CMR.ts:38]: So TS is now enforcing the existence of the arguments?

Comment (yjpa7145) [reply packages/cmr-client/src/CMR.ts:38]: It is, but that commented-out code definitely wasn't enforcing it before.

File: packages/cmr-client/src/CMR.ts [hunk 3]
```diff
@@ -219,20 +256,24 @@ class CMR {
...
     try {
       await validateUMMG(ummgMetadata, granuleId, this.provider);
 
       response = await got.put(
-        `${getUrl('ingest', this.provider)}granules/${granuleId}`,
+        `${getIngestUrl({ provider: this.provider })}granules/${granuleId}`,
         {
-          json: true,
-          body: ummgMetadata,
+          json: ummgMetadata,
```
Comment (markdboyd) [inline packages/cmr-client/src/CMR.ts:270]: These request properties certainly seem different, so I assume that is intentional?

File: packages/cmr-client/package.json (hunk 1 above)
Comment (yjpa7145) [reply packages/cmr-client/package.json:43]: I wanted to make sure that the version of the type matched the version of the package, but I guess it could probably be moved up.

File: packages/cmr-client/src/CMR.ts (hunk 3 above)
Comment (yjpa7145) [reply packages/cmr-client/src/CMR.ts:270]: Yes, that is a result of updating the version of `got`.

File: packages/cmr-client/src/CMRSearchConceptQueue.ts [hunk 4]
```diff
@@ -87,15 +109,19 @@ class CMRSearchConceptQueue {
...
     const results = await this.CMR.searchConcept(
       this.type,
       this.params,
       this.format,
       false
     );
     this.items = results;
-    this.params.page_num = (this.params.page_num) ? this.params.page_num + 1 : 1;
+
+    const paramsPageNum = this.params.get('page_num') ?? '0';
```
Comment (markdboyd) [inline packages/cmr-client/src/CMRSearchConceptQueue.ts:121]: Is `'0'` as a string intentional?

Comment (markdboyd) [reply packages/cmr-client/src/CMRSearchConceptQueue.ts:121]: I guess so based on the next line. Why are we maintaining it as a string?

File: packages/cmr-client/src/Utils.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import { parseStringPromise } from 'xml2js';
+
+export function parseXMLString(xmlString: string): Promise<unknown> {
```
Comment (markdboyd) [inline packages/cmr-client/src/Utils.ts:3]: Did we drop the unneeded promisify dependency?

File: packages/cmr-client/src/getUrl.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
+function getCmrHost(cmrEnvironment?: string): string {
```
Comment (markdboyd) [inline packages/cmr-client/src/getUrl.ts:1]: Much, much cleaner 👍

File: packages/cmr-client/src/getUrl.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  host,
+  cmrEnv = process.env.CMR_ENVIRONMENT,
+}: {
+  host?: string,
+  cmrEnv?: string,
+} = {}): string {
+  return `https://${host ?? getCmrHost(cmrEnv)}/search/`;
+}
+
+export function getTokenUrl({
```
Comment (markdboyd) [inline packages/cmr-client/src/getUrl.ts:36]: Clearly you removed getUrl, just want to make sure we're not still relying on it anywhere

File: packages/cmr-client/src/ingestConcept.ts [hunk 8]
```diff
@@ -25,27 +23,33 @@ const logDetails = {
...
-      `${getUrl('ingest', provider)}${type}s/${identifier}`,
+      `${getIngestUrl({ provider })}${type}s/${identifier}`,
       {
         body: xmlString,
         headers,
       }
     );
 
-    xmlObject = await parseXMLString(response.body);
+    const xmlErrorObject = <{errors?: {error: string}}>(await parseXMLString(response.body));
```
Comment (markdboyd) [inline packages/cmr-client/src/ingestConcept.ts:49]: It's not necessarily an error object unless it has `errors`? `xmlObject` seems like a better name

File: packages/cmr-client/tests/test-getUrl.js [hunk 9]
```diff
@@ -2,26 +2,26 @@
...
+test.serial('getSearchUrl returns value according to cmrEnvironment param', (t) => {
+  t.is(getSearchUrl({ cmrEnv: 'OPS' }), 'https://cmr.earthdata.nasa.gov/search/');
+  t.is(getSearchUrl({ cmrEnv: 'SIT' }), 'https://cmr.sit.earthdata.nasa.gov/search/');
+  t.is(getSearchUrl({ cmrEnv: 'UAT' }), 'https://cmr.uat.earthdata.nasa.gov/search/');
 });
 
 test.serial('getUrl pulls cmrEnvironment from environment variables', (t) => {
   process.env.CMR_ENVIRONMENT = 'SIT';
-  t.is(getUrl('search'), 'https://cmr.sit.earthdata.nasa.gov/search/');
+  t.is(getSearchUrl(), 'https://cmr.sit.earthdata.nasa.gov/search/');
```
Comment (markdboyd) [inline packages/cmr-client/tests/test-getUrl.js:15]: Should update the test name here to `getSearchUrl() ..`

File: packages/cmr-client/tests/test-getUrl.js [hunk 10]
```diff
@@ -2,26 +2,26 @@
...
 
   process.env.CMR_ENVIRONMENT = 'OPS';
-  t.is(getUrl('search'), 'https://cmr.earthdata.nasa.gov/search/');
+  t.is(getSearchUrl(), 'https://cmr.earthdata.nasa.gov/search/');
 });
 
 test.serial('getUrl uses cmrEnv from parameter over env variable', (t) => {
   process.env.CMR_ENVIRONMENT = 'SIT';
-  t.is(getUrl('search', null, 'OPS'), 'https://cmr.earthdata.nasa.gov/search/');
+  t.is(getSearchUrl({ cmrEnv: 'OPS' }), 'https://cmr.earthdata.nasa.gov/search/');
```
Comment (markdboyd) [inline packages/cmr-client/tests/test-getUrl.js:26]: Same: update the test name

File: packages/cmr-client/tests/test-searchConcept.js [hunk 11]
```diff
@@ -4,20 +4,22 @@ const got = require('got');
...
+process.env.CMR_ENVIRONMENT = 'SIT';
 
 const clientId = 'cumulus-test-client';
 
 test.serial('searchConcept request includes CMR client id', async (t) => {
   let request;
   const stub = sinon.stub(got, 'get').callsFake((_url, opt) => {
     request = { headers: opt.headers };
-    return { body: { feed: { entry: [] } }, headers: { 'cmr-hits': 0 } };
+    return { body: { feed: { entry: [] } }, headers: { 'cmr-hits': '0' } };
```
Comment (markdboyd) [inline packages/cmr-client/tests/test-searchConcept.js:17]: I probably missed this in the code, but why this change?

**Review by markdboyd — CHANGES_REQUESTED** [review]
Looks great! Several comments and a couple minor requested changes

File: packages/cmr-client/src/getUrl.ts (hunk 7 above)
Comment (yjpa7145) [reply packages/cmr-client/src/getUrl.ts:36]: I _thought_ I had, but I must not have committed that work or something. This commit removes the remaining references to `getUrl`: 8bc02f4f3

File: packages/cmr-client/src/ingestConcept.ts (hunk 8 above)
Comment (yjpa7145) [reply packages/cmr-client/src/ingestConcept.ts]: ea9797eb0

File: packages/cmr-client/tests/test-getUrl.js (hunk 9 above)
Comment (yjpa7145) [reply packages/cmr-client/tests/test-getUrl.js:15]: 8bc02f4f3

File: packages/cmr-client/tests/test-getUrl.js (hunk 10 above)
Comment (yjpa7145) [reply packages/cmr-client/tests/test-getUrl.js:26]: 8bc02f4f3

File: packages/cmr-client/tests/test-searchConcept.js (hunk 11 above)
Comment (yjpa7145) [reply packages/cmr-client/tests/test-searchConcept.js:17]: `URLSearchParams` only takes strings and null

**Comment by yjpa7145** [comment]
@markdboyd I think I've addressed all of the issues that you brought up. This is ready for a re-review.

**Review by markdboyd — APPROVED** [review]
